File: lib/wurk/debounce.rb
# frozen_string_literal: true

require_relative 'job_util'
require_relative 'keys'
require_relative 'lua'
require_relative 'pool_checkout'
require_relative 'unique'

module Wurk
  # Debounce — collapse a burst of enqueues into one job carrying the last
  # payload, fired after a period of quiet. A Wurk extra, not a Sidekiq
  # surface, and deliberately not a second flavour of {Wurk::Unique}: unique
  # *rejects* the duplicate and keeps the first, debounce *replaces* the
  # pending job and keeps the last.
  #
  # The classic use is a re-derivation triggered by something a user does
  # repeatedly — reindex a record, warm a cache, recompute a projection. Fifty
  # edits in a minute should produce one rebuild at the end, not fifty.
  #
  # Redis schema — one new key, plus an ordinary entry in the ZSET that already
  # exists:
  #
  #   debounce:<digest>   HASH   member (the pending job JSON), first (epoch)
  #   schedule            ZSET   the collapsed job, exactly as `perform_in`
  #                              would have written it
  #
  # Nothing reading `schedule` can tell a debounced job from any other
  # scheduled one — {Wurk::ScheduledSet}, the dashboard and a stock Sidekiq
  # poller all handle it unchanged. There is no parallel delayed structure.
  #
  # Identity is {Wurk::Unique}'s digest of `[class, queue, args]`, so a class
  # that already narrows its key with `sidekiq_unique_context` narrows its
  # debounce key the same way — that hook is the supported "collapse on a
  # subset of the args" escape hatch, and there is no second one.
  module Debounce
    # How long `debounce:<digest>` outlives the entry it points at.
    #
    # The key's only job is to remember which member to pull back out of
    # `schedule` when the burst is extended, so it must not expire while that
    # member is still pending — a burst that lost its key ZADDs a second entry
    # next to the first and stops collapsing. The window that matters is
    # therefore "fire time → poller promotes it", which the scheduler keeps at
    # roughly `average_scheduled_poll_interval` across the cluster however many
    # processes are running. Five minutes is far past that; overshooting costs
    # one ~100-byte key per idle debounce identity, and the script treats an
    # outlived key as a finished burst anyway.
    GRACE = 300

    # What one debounced enqueue did to the pending entry.
    #
    # `replaced` is the only part a caller cannot work out for itself: from
    # outside the script, extending a live burst and opening a new one look
    # identical. `at` is the epoch second the job will run — capped, so it is
    # not necessarily `now + wait`. `opened_at` is the burst's own start, the
    # clock `max_wait` is measured against.
    Outcome = Struct.new(:replaced, :at, :opened_at) do
      # True when this push displaced a still-pending sibling.
      def replaced? = replaced
    end

    class << self
      # @param job [Hash] a normalized job payload
      # @return [String] `debounce:<digest>` for that job's identity
      def key_for(job) = Keys.debounce(Unique.digest_for(job))

      # Collapse `job` into the pending entry for its key, or open a new burst.
      #
      # @param job [Hash] normalized payload; stored minus `at`/`enqueued_at`,
      #   the same bytes {Wurk::Client#push_scheduled} would have written
      # @param wait [Numeric] seconds of quiet before the job fires; every
      #   further enqueue of the same key pushes this out again
      # @param max_wait [Numeric, nil] hard cap measured from the *first*
      #   enqueue of the burst. nil leaves the job able to starve — a key
      #   re-enqueued faster than `wait` never fires — so callers that cannot
      #   rule that out should always set it.
      # @param pool [#with, nil] defaults to this process's pool
      # @return [Outcome]
      def schedule(job, wait:, max_wait: nil, pool: nil)
        seconds = positive_seconds!(wait, 'wait')
        cap = max_wait.nil? ? 0.0 : positive_seconds!(max_wait, 'max_wait')
        reject_shorter_cap!(seconds, cap)

        argv = [seconds, cap, JobUtil.scheduled_member(job), GRACE]
        # Replay-safe: a re-run finds the member it just wrote, ZREMs it and
        # ZADDs it back, so a pool retry after a lost reply converges on one
        # entry rather than a duplicate — the whole point of doing this in a
        # script rather than a pipeline.
        #
        # The replay is not byte-identical: it recomputes `now`, so the fire
        # time moves out by the retry's own latency (and `first` is carried
        # forward, so `max_wait` still caps where it lands). That is the claim
        # worth making anyway. A lost reply means the write probably applied,
        # and refusing the replay only converts a bounded shift into an
        # exception out of `perform_async` for a job that is already scheduled
        # — whose caller then re-enqueues and shifts the fire time regardless.
        raw = with_pool(pool, idempotent: true) do |conn|
          Lua::Loader.eval_cached(conn, :debounce, keys: [key_for(job), Keys::SCHEDULE], argv: argv)
        end
        Outcome.new(raw[0].to_i == 1, raw[1].to_f, raw[2].to_f)
      end

      private

      def positive_seconds!(value, name)
        raise ArgumentError, "debounce #{name} must be a positive number of seconds: #{value.inspect}" \
          unless JobUtil.positive_seconds?(value)

        value.to_f
      end

      # A cap below the quiet period fires every burst at `first + max_wait`
      # regardless of traffic, which is a throttle wearing a debounce's name.
      # Nobody writes that on purpose, so it fails where it was declared.
      def reject_shorter_cap!(wait, cap)
        return if cap.zero? || cap >= wait

        raise ArgumentError, "debounce max_wait (#{cap}) must be at least wait (#{wait})"
      end

      def with_pool(pool, idempotent: false, &)
        pool ? PoolCheckout.with(pool, idempotent, &) : Wurk.redis(idempotent:, &)
      end
    end
  end
end

File: lib/wurk/lua/debounce.lua
-- Debounce: collapse a burst of enqueues of the same key down to one
-- scheduled job carrying the LAST payload, fired after `wait` seconds of
-- quiet.
--
-- One script, not read-then-write. Reading the pending member, pulling it back
-- out of `schedule`, writing the replacement and restamping the burst all have
-- to land without another producer interleaving: two processes debouncing the
-- same key at the same instant would otherwise each see "nothing pending" and
-- each ZADD, which is exactly the one thing a collapse policy must not do.
--
-- ARGV[2] is the starvation guard, and it is why `first` is stored at all. A
-- key re-enqueued faster than `wait` pushes its own fire time forward on every
-- push and never runs — the classic debounce bug. `first` pins when the burst
-- opened, and the fire time is capped at first + max_wait, so a permanently
-- busy key still fires on a fixed cadence instead of never.
--
-- What lands in `schedule` is an ordinary scheduled job: score = epoch
-- seconds, member = the caller's job JSON, no side structures. Wurk's
-- ScheduledSet, the dashboard and a stock Sidekiq poller all read it as any
-- other `perform_in` — a debounced job is not a distinguishable kind of job.
--
-- The clock is Redis's, not the caller's. Producers spread across machines
-- with skewed clocks then debounce against one window, and a host whose clock
-- runs fast cannot drag the max_wait cap forward for everyone else.
--
-- KEYS[1] = debounce:<digest>
-- KEYS[2] = schedule
-- ARGV[1] = wait seconds (float ok)
-- ARGV[2] = max_wait seconds measured from the first enqueue; 0 = uncapped
-- ARGV[3] = job JSON, already stripped of `at`/`enqueued_at` (the ZSET member)
-- ARGV[4] = grace seconds the key outlives its own entry by
-- Returns {replaced, fire_at, burst_started_at}; `replaced` is 1 when this
-- call displaced a still-pending sibling, 0 when it opened a new burst.
local dkey, zkey = KEYS[1], KEYS[2]
local wait = tonumber(ARGV[1])
local max_wait = tonumber(ARGV[2])
local member = ARGV[3]
local grace = tonumber(ARGV[4])

local t = redis.call('TIME')
local now = tonumber(t[1]) + (tonumber(t[2]) / 1000000)

local replaced = 0
local first = now
local prev = redis.call('HMGET', dkey, 'member', 'first')
-- The ZREM doubles as the liveness test. A recorded member that is no longer
-- in `schedule` has already been promoted by the poller (or deleted from the
-- dashboard): its burst is over, so this enqueue opens a new one. Carrying the
-- old `first` forward instead would cap a fresh burst against a deadline that
-- has already passed and fire it immediately, forever.
if prev[1] and redis.call('ZREM', zkey, prev[1]) == 1 then
  replaced = 1
  first = tonumber(prev[2]) or now
end

local at = now + wait
if max_wait > 0 then
  local cap = first + max_wait
  if at > cap then at = cap end
end

-- Fixed %.6f rather than letting Redis format the double: TIME is
-- microsecond-resolution, so six decimals is lossless here, and the caller
-- gets back a score it can compare byte-for-byte with what is in the ZSET.
local score = string.format('%.6f', at)
local started = string.format('%.6f', first)
redis.call('ZADD', zkey, score, member)
redis.call('HSET', dkey, 'member', member, 'first', started)
-- Whole seconds, rounded up, floored at 1: the key must outlive the entry it
-- guards. If it expires first, the next enqueue sees no pending member and
-- ZADDs a second entry alongside the one still sitting in `schedule` — the
-- grace covers however long the poller takes to promote a due job.
redis.call('EXPIRE', dkey, math.max(1, math.ceil((at - now) + grace)))

return { replaced, score, started }
